src/objects.c

```diff
@@ -440,6 +440,9 @@ void p11prov_obj_free(P11PROV_OBJ *obj)
     if (obj == NULL) {
         return;
     }
+
+    p11prov_obj_free(obj->pub_key_obj);
+
```

Comment:
This is wrong, all freeing MUST happen *after* the refcounting reaches 0.
Move it after the following block and you'll solve most of your segfaults.